src/transformers/trainer_callback.py

```diff
@@ -475,3 +476,62 @@ def on_log(self, args, state, control, logs=None, **kwargs):
         _ = logs.pop("total_flos", None)
         if state.is_local_process_zero:
             print(logs)
+
+
+class EarlyStoppingCallback(TrainerCallback):
+    """
+    An early stopping callback
+
+    Args:
+       early_stopping_patience (:obj:`int`):
+            Use with :obj:`metric_for_best_model` to stop training when the specified metric worsens for
+            :obj:`early_stopping_patience` evaluation calls.
+       early_stopping_threshold(:obj:`float`, `optional`):
+            Use with TrainingAruments :obj:`metric_for_best_model` and :obj:`early_stopping_patience` to denote how
+            much the specified metric must improve to satisfy early stopping conditions. `
+
+    This callback depends on :class:`~transformers.TrainingArguments` argument `load_best_model_at_end` functionality
+    to set best_metric in :class:`~transformers.TrainerState`.
+    """
+
+    def __init__(self, early_stopping_patience: int = 1, early_stopping_threshold: Optional[float] = 0.0):
+        self.early_stopping_patience = early_stopping_patience
+        self.early_stopping_threshold = early_stopping_threshold
+        # early_stopping_patience_counter denotes the number of times validation metrics failed to improve.
+        self.early_stopping_patience_counter = 0
+
+    def check_metric_value(self, args, state, control, metric_value):
+        # best_metric is set by code for load_best_model
+        operator = np.greater if args.greater_is_better else np.less
+        if state.best_metric is None or (
+            operator(metric_value, state.best_metric)
+            and abs(metric_value - state.best_metric) > self.early_stopping_threshold
+        ):
+            self.early_stopping_patience_counter = 0
+        else:
+            self.early_stopping_patience_counter += 1
+
+    def on_train_begin(self, args, state, control, **kwargs):
+        assert args.load_best_model_at_end, "EarlyStoppingCallback requires load_best_model_at_end = True"
```

Comment:
I still don't understand why this line is necessary? I feel we should be able to use this callback without the option `load_best_model_at_end`? The other sanity checks are perfectly ok.

Reply:
This is necessary because we need  `trainer._save_checkpoint` to calculate the best_metric

Reply:
This is necessary because we require `control.should_save=True` for [_save_checkpoint](https://github.com/huggingface/transformers/blob/63e91f5fde64563bdf19276d3eca9b61fc8310c3/src/transformers/trainer.py#L847) to update the best metric. Should I move the best metric calculation into its own function and place it in the `should_evaluate` [block](https://github.com/huggingface/transformers/blob/63e91f5fde64563bdf19276d3eca9b61fc8310c3/src/transformers/trainer.py#L843)?

Reply:
I agree that it's not fully intuitive to need load_best_model_at_end, but it makes sense to me because if we don't load the best model early stopping will stop us, but the model we receive back from training will not be the model early stopping deemed best.

Reply:
Ok let's leave it as is for now then, and we will re-evaluate if some users complain!

Reply:
Saw this issue while debugging something. It doesn't seem intuitive how these two are related, so can we please do what @cbrochtrup suggested above?